lib/absinthe/middleware.ex

```diff
@@ -0,0 +1,264 @@
+defmodule Absinthe.Middleware do
+  @moduledoc """
+  Middleware enables custom resolution behaviour on a field.
+
+  All resolution happens through middleware. Even `resolve` functions are middleware,
+  as the `resolve` macro is just
+
+  ```
+  quote do
+    middleware Absinthe.Resolution, unquote(function_ast)
+  end
+  ```
+
+  Resolution happens by reducing a list of middleware onto an `%Absinthe.Resolution{}`
+  struct.
+
+  ## Example
+
+  ```
+  defmodule MyApp.Web.Authentication do
+    @behaviour Absinthe.Middleware
+
+    def call(resolution, _opts) do
+      case resolution.context do
+        %{current_user: _} ->
+          resolution
+        _ ->
+          resolution
+          |> Absinthe.Resolution.put_result({:error, "unauthenticated"})
+      end
+    end
+  end
+  ```
+
+  By specifying `@behaviour Absinthe.Middleware` the compiler will ensure that
+  we provide a `def call` callback. This function takes an `%Absinthe.Resolution{}`
+  struct and will also need to return one such struct.
+
+  On that struct there is a `context` key which holds the absinthe context. This
+  is generally where things like the current user are placed. For more information
+  on how the current user ends up in the context please see our full authentication
+  guide on the website.
+
+  Our `def call` function simply checks the context to see if there is a current
+  user. If there is, we pass the resolution onward. If there is not, we update
+  the resolution state to `:halt` and place an error result.
```

Comment:
Here it says that the resolution state is updated to `:halt`. Line 29 says `:error`. Is the halting a byproduct state of `put_result` an `:error` or should one of these be updated?